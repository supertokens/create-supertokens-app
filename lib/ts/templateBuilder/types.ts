// types/config.ts
export type ConfigType =
    | "all_auth"
    | "emailpassword"
    | "multifactorauth"
    | "multitenancy"
    | "passwordless"
    | "thirdparty"
    | "thirdpartyemailpassword"
    | "thirdpartypasswordless"
    | "webauthn";

export type IndividualRecipe =
    | "emailPassword"
    | "thirdParty"
    | "passwordless"
    | "session"
    | "dashboard"
    | "userRoles"
    | "multiFactorAuth"
    | "accountLinking"
    | "emailVerification"
    | "totp"
    | "multitenancy"
    | "webauthn";

export type FrontendFramework = "react" | "vue" | "angular" | "solid";

export type Language = "ts" | "py" | "go";

export type AppInfo = {
    appName: string;
    apiDomain: string;
    websiteDomain: string;
    apiBasePath: string;
    websiteBasePath: string;
    connectionURI: string;
};

export type OAuthProvider = {
    id: string;
    clientId: string;
    clientSecret: string;
    additionalConfig?: Record<string, any>;
};
