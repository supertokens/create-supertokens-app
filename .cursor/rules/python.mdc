---
description: 
globs: 
alwaysApply: false
---
## Python

| factor          | py imports                                                                                                                         | py inits                                                                                                                                                   |
| --------------- | ---------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `emailpassword` | from supertokens_python.recipe import emailpassword<br>                                                                            | emailpassword.init()                                                                                                                                       |
| `thirdparty`    | from supertokens_python.recipe import thirdparty                                                                                   | thirdparty.init(<br>        sign_in_and_up_feature=thirdparty.SignInAndUpFeature(providers=[<br>            // per-provider config<br>        ])<br>    ), |
| `link_email`    | from supertokens_python.recipe import paswordless<br><br>from supertokens_python.recipe.passwordless import ContactEmailConfig     | passwordless.init(<br>        flow_type="MAGIC_LINK",<br>        contact_config=ContactEmailConfig(),<br>    ),                                            |
| `link_phone`    | from supertokens_python.recipe import paswordless<br><br>from supertokens_python.recipe.passwordless import ContactPhoneConfig     | passwordless.init(<br>        flow_type="MAGIC_LINK",<br>        contact_config=ContactPhoneConfig(),<br>    ),                                            |
| `otp_phone`     | from supertokens_python.recipe import paswordless<br><br>from supertokens_python.recipe.passwordless import ContactPhoneConfig     | passwordless.init(<br>        flow_type="USER_INPUT_CODE",<br>        contact_config=ContactPhoneConfig(),<br>    ),                                       |
| `otp_email`     | from supertokens_python.recipe import <br>paswordless<br><br>from supertokens_python.recipe.passwordless import ContactEmailConfig | passwordless.init(<br>        flow_type="USER_INPUT_CODE",<br>        contact_config=ContactEmailConfig(),<br>    ),                                       |
| `totp`          | from supertokens_python.recipe import <br>totp                                                                                     | totp.init(),                                                                                                                                               |

**Notes**: 
- session, dashboard, userroles also imported for all factors, in addition to their own config.
- if both link_email and link_phone are present as factors, the contact_config in init becomes ContactEmailOrPhoneConfig().
- if both otp_email and otp_phone are present as factors, the flowType in init becomes "USER_INPUT_CODE_AND_MAGIC_LINK".
- If a second factor is present, always include the multifactorauth recipe.
- For any multifactor recipes, also include accountlinking and emailverfiation.
- Second factors specific behavior always go in the override function of the MultiFactorAuth recipe!


### Examples

#### All auth

```python
from supertokens_python import init, InputAppInfo, SupertokensConfig
from supertokens_python.recipe import emailpassword, thirdparty, passwordless, session, dashboard
from supertokens_python.recipe.passwordless import ContactEmailOrPhoneConfig
from supertokens_python.recipe.thirdparty.provider import ProviderInput, ProviderConfig, ProviderClientConfig
from supertokens_python import (
    InputAppInfo,
    SupertokensConfig,
)

# this is the location of the SuperTokens core.
supertokens_config = SupertokensConfig(
    connection_uri="https://try.supertokens.com")

app_info = InputAppInfo(
    app_name="Supertokens",
    api_domain="http://localhost:3001",
    website_domain="http://localhost:3000",
)

# recipeList contains all the modules that you want to
# use from SuperTokens. See the full list here: https://supertokens.com/docs/guides
recipe_list = [
    session.init(),
    emailpassword.init(),
    passwordless.init(
        flow_type="USER_INPUT_CODE_AND_MAGIC_LINK",
        contact_config=ContactEmailOrPhoneConfig(),
    ),
    thirdparty.init(
        sign_in_and_up_feature=thirdparty.SignInAndUpFeature(providers=[
            ProviderInput(
                config=ProviderConfig(
                    third_party_id="google",
                    clients=[
                        ProviderClientConfig(
                            client_id='1060725074195-kmeum4crr01uirfl2op9kd5acmi9jutn.apps.googleusercontent.com',
                            client_secret='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'
                        ),
                    ],
                ),
            ),
            ProviderInput(
                config=ProviderConfig(
                    third_party_id="github",
                    clients=[
                        ProviderClientConfig(
                            client_id='467101b197249757c71f',
                            client_secret='e97051221f4b6426e8fe8d51486396703012f5bd'
                        ),
                    ],
                ),
            ),
            ProviderInput(
                config=ProviderConfig(
                    third_party_id="apple",
                    clients=[
                        ProviderClientConfig(
                            client_id="4398792-io.supertokens.example.service",
                            additional_config={
                                "keyId": "7M48Y4RYDL",
                                "privateKey": "-----BEGIN PRIVATE KEY-----\nMIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgu8gXs+XYkqXD6Ala9Sf/iJXzhbwcoG5dMh1OonpdJUmgCgYIKoZIzj0DAQehRANCAASfrvlFbFCYqn3I2zeknYXLwtH30JuOKestDbSfZYxZNMqhF/OzdZFTV0zc5u5s3eN+oCWbnvl0hM+9IW0UlkdA\n-----END PRIVATE KEY-----",
                                "teamId": "YWQCXGJRJL",
                            },
                        ),
                    ],
                ),
            ),
            ProviderInput(
                config=ProviderConfig(
                    third_party_id="twitter",
                    clients=[
                        ProviderClientConfig(
                            client_id='4398792-WXpqVXRiazdRMGNJdEZIa3RVQXc6MTpjaQ',
                            client_secret='BivMbtwmcygbRLNQ0zk45yxvW246tnYnTFFq-LH39NwZMxFpdC'
                        ),
                    ],
                ),
            ),
        ])
    ),
    dashboard.init()
]
```

#### Multi-factor auth 

```python
from supertokens_python.recipe import (
    session,
    thirdparty,
    emailpassword,
    dashboard,
    passwordless,
    multifactorauth,
    emailverification,
    accountlinking,
    totp,
)
from supertokens_python.recipe.accountlinking.types import (
    AccountInfoWithRecipeIdAndUserId,
    ShouldAutomaticallyLink,
    ShouldNotAutomaticallyLink,
)
from supertokens_python.recipe.multifactorauth.types import OverrideConfig
from supertokens_python.recipe.passwordless import ContactEmailOrPhoneConfig
from supertokens_python.recipe.multifactorauth.interfaces import (
    RecipeInterface as MFARecipeInterface,
)
from supertokens_python.recipe.multifactorauth.types import (
    MFARequirementList,
    FactorIds,
)
from supertokens_python.recipe.session import SessionContainer
from supertokens_python.recipe.thirdparty.provider import (
    ProviderInput,
    ProviderConfig,
    ProviderClientConfig,
)
from supertokens_python import (
    InputAppInfo,
    SupertokensConfig,
)
from supertokens_python.types import User
from typing import Optional, Dict, Any, Callable, Awaitable, List, Union


# this is the location of the SuperTokens core.
supertokens_config = SupertokensConfig(connection_uri="https://try.supertokens.com")

app_info = InputAppInfo(
    app_name="Supertokens",
    api_domain="http://localhost:3001",
    website_domain="http://localhost:3000",
)


async def should_do_automatic_account_linking(
    new_account_info: AccountInfoWithRecipeIdAndUserId,
    user: Optional[User],
    session: Optional[SessionContainer],
    tenant_id: str,
    user_context: Dict[str, Any],
) -> Union[ShouldNotAutomaticallyLink, ShouldAutomaticallyLink]:
    return ShouldAutomaticallyLink(should_require_verification=True)


def override_multifactor_functions(original_implementation: MFARecipeInterface):
    # Override the MFA requirements
    async def get_mfa_requirements_for_auth(
        tenant_id: str,
        access_token_payload: Dict[str, Any],
        completed_factors: Dict[str, int],
        user: Callable[[], Awaitable[User]],
        factors_set_up_for_user: Callable[[], Awaitable[List[str]]],
        required_secondary_factors_for_user: Callable[[], Awaitable[List[str]]],
        required_secondary_factors_for_tenant: Callable[[], Awaitable[List[str]]],
        user_context: Dict[str, Any],
    ) -> MFARequirementList:
        return [
            {
                "oneOf": [
                    FactorIds.TOTP,
                    FactorIds.OTP_EMAIL,
                    FactorIds.OTP_PHONE,
                ],
            }
        ]

    # Patch the original implementation to use our overridden function
    original_implementation.get_mfa_requirements_for_auth = (
        get_mfa_requirements_for_auth
    )
    return original_implementation


# recipeList contains all the modules that you want to
# use from SuperTokens. See the full list here: https://supertokens.com/docs/guides
recipe_list = [
    emailpassword.init(),
    thirdparty.init(
        sign_in_and_up_feature=thirdparty.SignInAndUpFeature(
            providers=[
                # We have provided you with development keys which you can use for testing.
                # IMPORTANT: Please replace them with your own OAuth keys for production use.
                ProviderInput(
                    config=ProviderConfig(
                        third_party_id="google",
                        clients=[
                            ProviderClientConfig(
                                client_id="1060725074195-kmeum4crr01uirfl2op9kd5acmi9jutn.apps.googleusercontent.com",
                                client_secret="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",
                            ),
                        ],
                    ),
                ),
                ProviderInput(
                    config=ProviderConfig(
                        third_party_id="github",
                        clients=[
                            ProviderClientConfig(
                                client_id="467101b197249757c71f",
                                client_secret="e97051221f4b6426e8fe8d51486396703012f5bd",
                            ),
                        ],
                    ),
                ),
                ProviderInput(
                    config=ProviderConfig(
                        third_party_id="apple",
                        clients=[
                            ProviderClientConfig(
                                client_id="4398792-io.supertokens.example.service",
                                additional_config={
                                    "keyId": "7M48Y4RYDL",
                                    "privateKey": "-----BEGIN PRIVATE KEY-----\nMIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgu8gXs+XYkqXD6Ala9Sf/iJXzhbwcoG5dMh1OonpdJUmgCgYIKoZIzj0DAQehRANCAASfrvlFbFCYqn3I2zeknYXLwtH30JuOKestDbSfZYxZNMqhF/OzdZFTV0zc5u5s3eN+oCWbnvl0hM+9IW0UlkdA\n-----END PRIVATE KEY-----",
                                    "teamId": "YWQCXGJRJL",
                                },
                            ),
                        ],
                    ),
                ),
                ProviderInput(
                    config=ProviderConfig(
                        third_party_id="twitter",
                        clients=[
                            ProviderClientConfig(
                                client_id="4398792-WXpqVXRiazdRMGNJdEZIa3RVQXc6MTpjaQ",
                                client_secret="BivMbtwmcygbRLNQ0zk45yxvW246tnYnTFFq-LH39NwZMxFpdC",
                            ),
                        ],
                    ),
                ),
            ]
        )
    ),
    passwordless.init(
        flow_type="USER_INPUT_CODE_AND_MAGIC_LINK",
        contact_config=ContactEmailOrPhoneConfig(),
    ),
    emailverification.init(mode="REQUIRED"),
    accountlinking.init(
        should_do_automatic_account_linking=should_do_automatic_account_linking
    ),
    multifactorauth.init(
        first_factors=["thirdparty", "emailpassword"],
        override=OverrideConfig(
            functions=override_multifactor_functions,
        ),
    ),
    totp.init(),
    session.init(),
    dashboard.init(),
]
```

#### Multitenancy

```python
from supertokens_python.recipe import session, thirdparty, emailpassword, dashboard, passwordless
from supertokens_python.recipe.passwordless import ContactEmailOrPhoneConfig
from supertokens_python import (
    InputAppInfo,
    SupertokensConfig,
)

# this is the location of the SuperTokens core.
supertokens_config = SupertokensConfig(
    connection_uri="https://try.supertokens.com")

app_info = InputAppInfo(
    app_name="Supertokens",
    api_domain="http://localhost:3001",
    website_domain="http://localhost:3000",
)

# recipeList contains all the modules that you want to
# use from SuperTokens. See the full list here: https://supertokens.com/docs/guides
recipe_list = [
    session.init(),
    thirdparty.init(),
    emailpassword.init(),
    passwordless.init(
        flow_type="USER_INPUT_CODE_AND_MAGIC_LINK",
        contact_config=ContactEmailOrPhoneConfig(),
    ),
    dashboard.init()
]
```

These rules and examples apply only to [py.ts](mdc:boilerplate/backend/shared/python/py.ts) - and nowhere outside of it.



