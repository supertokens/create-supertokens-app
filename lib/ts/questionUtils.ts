import os from "os";
import { Answers, QuestionOption, RecipeQuestionOption, UserFlags } from "./types.js";

export function getPythonRunScripts(): string[] {
    if (os.platform() === "win32") {
        return [
            "pip install virtualenv",
            "python -m virtualenv venv",
            ".\\\\venv\\\\Scripts\\\\activate.bat",
            "pip install -r requirements.txt",
            "python app.py",
        ];
    }
    return [
        "pip install virtualenv",
        "virtualenv venv",
        "chmod +x venv/bin/activate",
        ". venv/bin/activate",
        "pip install -r requirements.txt",
        "python app.py",
    ];
}

export function getDjangoPythonRunScripts(): string[] {
    if (os.platform() === "win32") {
        return [
            "pip install virtualenv",
            "python -m virtualenv venv",
            ".\\\\venv\\\\Scripts\\\\activate.bat",
            "pip install -r requirements.txt",
            "python manage.py runserver",
        ];
    }
    return [
        "pip install virtualenv",
        "virtualenv venv",
        "chmod +x venv/bin/activate",
        ". venv/bin/activate",
        "pip install -r requirements.txt",
        "python manage.py runserver",
    ];
}

// Converts the options array we declare to a format inquirer can use
export function mapOptionsToChoices(options: QuestionOption[] | RecipeQuestionOption[]) {
    return options
        .filter((i) => i.shouldDisplay !== false)
        .map((option) => {
            return {
                name: option.displayName,
                value: option.value,
            };
        });
}

/**
 * This modifies the answers hash to use the correct python framework id if the user selects python
 * for their backend
 */
export function modifyAnswersForPythonFrameworks(answers: Answers) {
    let _answers = answers;

    if (answers.backend === "python") {
        _answers.backend = answers.backendPython;
    }

    return _answers;
}

export function modifyAnswersForNodeJSFrameworks(answers: Answers) {
    let _answers = answers;
    if (answers.backend === "node") {
        _answers.backend = answers.backendNodeJS;
    }
    return _answers;
}

/**
 * Decides whether the user should be prompted to select their backend.This question is skipped for full stack frameworks
 */
export function shouldSkipBackendQuestion(answers: Answers, userFlags: UserFlags): boolean {
    let frontEndInFlags = userFlags.frontend;

    if (frontEndInFlags !== undefined) {
        // Priority goes to flags
        return (
            frontEndInFlags.startsWith("next") ||
            frontEndInFlags === "capacitor" ||
            frontEndInFlags === "remix" ||
            frontEndInFlags === "astro-react" ||
            frontEndInFlags === "sveltekit" ||
            frontEndInFlags === "nuxtjs"
        );
    }

    return (
        (answers.frontend !== undefined && answers.frontend.startsWith("next")) ||
        answers.frontend === "capacitor" ||
        answers.frontend === "remix" ||
        answers.frontend === "astro-react" ||
        answers.frontend === "sveltekit" ||
        answers.frontend === "nuxtjs"
    );
}
