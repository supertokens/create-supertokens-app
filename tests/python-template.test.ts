import { describe, it, expect } from "vitest";
import { generatePythonTemplate } from "../boilerplate/backend/shared/python/py"; // Adjust path
import { UserFlags } from "../lib/ts/types"; // Adjust path

describe("Python Template Generation", () => {
    it("should generate correct config for EmailPassword only (FastAPI)", () => {
        const args = {
            configType: "emailpassword" as const,
            userArguments: { manager: "npm" } as UserFlags,
            framework: "fastapi",
        };
        const result = generatePythonTemplate(args);
        expect(result).toMatchSnapshot();
    });

    it("should generate correct config for ThirdParty only (Flask)", () => {
        const args = {
            configType: "thirdparty" as const,
            userArguments: { manager: "npm" } as UserFlags,
            framework: "flask",
        };
        const result = generatePythonTemplate(args);
        expect(result).toMatchSnapshot();
    });

    it("should generate correct config for ThirdParty + EmailPassword + TOTP MFA (DRF)", () => {
        const args = {
            configType: "thirdpartyemailpassword" as const,
            userArguments: {
                manager: "npm" as const,
                firstfactors: ["emailpassword", "thirdparty"],
                secondfactors: ["totp"],
            } as UserFlags,
            framework: "drf",
        };
        const result = generatePythonTemplate(args);
        expect(result).toMatchSnapshot();
    });

    it("should generate correct config for EmailPassword + OTP/Link Email/Phone (FastAPI)", () => {
        const args = {
            configType: "emailpassword" as const, // configType might be inferred differently, but test factors
            userArguments: {
                manager: "npm" as const,
                firstfactors: ["emailpassword", "otp-email", "link-email", "otp-phone", "link-phone"],
            } as UserFlags,
            framework: "fastapi",
        };
        const result = generatePythonTemplate(args);
        // Check specific parts for passwordless config
        expect(result).toContain("contact_config=ContactEmailOrPhoneConfig()");
        expect(result).toContain('flow_type="USER_INPUT_CODE_AND_MAGIC_LINK"');
        expect(result).toMatchSnapshot();
    });

    // TODO: Add more specific test cases for different factor combinations and frameworks
});
