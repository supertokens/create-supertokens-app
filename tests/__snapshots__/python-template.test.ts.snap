// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Python Template Generation > should generate correct config for EmailPassword + OTP/Link Email/Phone (FastAPI) 1`] = `
"from supertokens_python.recipe import session
from supertokens_python.recipe import dashboard
from supertokens_python.recipe import userroles
from supertokens_python.recipe import emailpassword
from supertokens_python.recipe import passwordless
from supertokens_python.recipe.passwordless import ContactEmailOrPhoneConfig



from supertokens_python import init, InputAppInfo, SupertokensConfig
import os

def get_api_domain() -> str:
    api_port = str(3001)
    api_url = f"http://localhost:{api_port}"
    return api_url

def get_website_domain() -> str:
    website_port = str(3000)
    website_url = f"http://localhost:{website_port}"
    return website_url

supertokens_config = SupertokensConfig(
    connection_uri="https://try.supertokens.com"
)

app_info = InputAppInfo(
    app_name="SuperTokens Demo App",
    api_domain=get_api_domain(),
    website_domain=get_website_domain(),
    api_base_path="/auth",
    website_base_path="/auth"
)

recipe_list = [
    session.init(),
    dashboard.init(),
    userroles.init(),
    emailpassword.init(),
    passwordless.init(
        flow_type="USER_INPUT_CODE_AND_MAGIC_LINK",
        contact_config=ContactEmailOrPhoneConfig()
    )
]

init(
    supertokens_config=supertokens_config,
    app_info=app_info,
    framework="fastapi",
    recipe_list=recipe_list,
    mode="asgi",
    telemetry=False
)
"
`;

exports[`Python Template Generation > should generate correct config for EmailPassword only (FastAPI) 1`] = `
"from supertokens_python.recipe import session
from supertokens_python.recipe import dashboard
from supertokens_python.recipe import userroles
from supertokens_python.recipe import emailpassword



from supertokens_python import init, InputAppInfo, SupertokensConfig
import os

def get_api_domain() -> str:
    api_port = str(3001)
    api_url = f"http://localhost:{api_port}"
    return api_url

def get_website_domain() -> str:
    website_port = str(3000)
    website_url = f"http://localhost:{website_port}"
    return website_url

supertokens_config = SupertokensConfig(
    connection_uri="https://try.supertokens.com"
)

app_info = InputAppInfo(
    app_name="SuperTokens Demo App",
    api_domain=get_api_domain(),
    website_domain=get_website_domain(),
    api_base_path="/auth",
    website_base_path="/auth"
)

recipe_list = [
    session.init(),
    dashboard.init(),
    userroles.init(),
    emailpassword.init()
]

init(
    supertokens_config=supertokens_config,
    app_info=app_info,
    framework="fastapi",
    recipe_list=recipe_list,
    mode="asgi",
    telemetry=False
)
"
`;

exports[`Python Template Generation > should generate correct config for ThirdParty + EmailPassword + TOTP MFA (DRF) 1`] = `
"from supertokens_python.recipe import session
from supertokens_python.recipe import dashboard
from supertokens_python.recipe import userroles
from supertokens_python.recipe import emailpassword
from supertokens_python.recipe import thirdparty
from supertokens_python.recipe.thirdparty.provider import ProviderInput, ProviderConfig, ProviderClientConfig
from supertokens_python.recipe import multifactorauth
from supertokens_python.recipe import totp
from supertokens_python.recipe import emailverification
from supertokens_python.recipe import accountlinking
from supertokens_python.recipe.accountlinking.types import AccountInfoWithRecipeIdAndUserId, ShouldAutomaticallyLink, ShouldNotAutomaticallyLink
from supertokens_python.recipe.session.interfaces import SessionContainer
from supertokens_python.types import User
from supertokens_python.recipe.multifactorauth.interfaces import RecipeInterface as MFARecipeInterface
from supertokens_python.recipe.multifactorauth.types import MFARequirementList, FactorIds
from typing import Optional, Dict, Any, Union, Callable, Awaitable, List

def override_multifactor_functions(original_implementation: MFARecipeInterface):
    async def get_mfa_requirements_for_auth(
        tenant_id: str,
        access_token_payload: dict,
        completed_factors: dict,
        user,
        factors_set_up_for_user,
        required_secondary_factors_for_user,
        required_secondary_factors_for_tenant,
        user_context: dict,
    ) -> MFARequirementList:
        return [
            {
                "oneOf": [
                    FactorIds.TOTP
                ],
            }
        ]

    async def get_required_secondary_factors_for_user(
        tenant_id: str,
        user_id: str,
        user_context: dict,
    ) -> list:
        return [FactorIds.TOTP]

    original_implementation.get_mfa_requirements_for_auth = get_mfa_requirements_for_auth
    original_implementation.get_required_secondary_factors_for_user = get_required_secondary_factors_for_user
    return original_implementation


async def async_should_do_linking(
    new_account_info: AccountInfoWithRecipeIdAndUserId,
    user: Optional[User],
    session: Optional[SessionContainer],
    tenant_id: str,
    user_context: Dict[str, Any],
) -> Union[ShouldNotAutomaticallyLink, ShouldAutomaticallyLink]:
    return ShouldAutomaticallyLink(should_require_verification=False)



from supertokens_python import init, InputAppInfo, SupertokensConfig
import os

def get_api_domain() -> str:
    api_port = str(3001)
    api_url = f"http://localhost:{api_port}"
    return api_url

def get_website_domain() -> str:
    website_port = str(3000)
    website_url = f"http://localhost:{website_port}"
    return website_url

supertokens_config = SupertokensConfig(
    connection_uri="https://try.supertokens.com"
)

app_info = InputAppInfo(
    app_name="SuperTokens Demo App",
    api_domain=get_api_domain(),
    website_domain=get_website_domain(),
    api_base_path="/auth",
    website_base_path="/auth"
)

recipe_list = [
    session.init(),
    dashboard.init(),
    userroles.init(),
    emailpassword.init(),
    thirdparty.init(
        sign_in_and_up_feature=thirdparty.SignInAndUpFeature(
            providers=[
                ProviderInput(
                    config=ProviderConfig(
                        third_party_id="google",
                        clients=[
                            ProviderClientConfig(
                                client_id="1060725074195-kmeum4crr01uirfl2op9kd5acmi9jutn.apps.googleusercontent.com",
                                client_secret="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"
                            )
                        ]
                    )
                ),
                ProviderInput(
                    config=ProviderConfig(
                        third_party_id="github",
                        clients=[
                            ProviderClientConfig(
                                client_id="467101b197249757c71f",
                                client_secret="e97051221f4b6426e8fe8d51486396703012f5bd"
                            )
                        ]
                    )
                ),
                ProviderInput(
                    config=ProviderConfig(
                        third_party_id="apple",
                        clients=[
                            ProviderClientConfig(
                                client_id="4398792-io.supertokens.example.service",
                                client_secret="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",
                                additional_config={
          "keyId": "7M48Y4RYDL",
          "privateKey": "-----BEGIN PRIVATE KEY-----\\nMIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgu8gXs+XYkqXD6Ala9Sf/iJXzhbwcoG5dMh1OonpdJUmgCgYIKoZIzj0DAQehRANCAASfrvlFbFCYqn3I2zeknYXLwtH30JuOKestDbSfZYxZNMqhF/OzdZFTV0zc5u5s3eN+oCWbnvl0hM+9IW0UlkdA\\n-----END PRIVATE KEY-----",
          "teamId": "YWQCXGJRJL"
}
                            )
                        ]
                    )
                ),
                ProviderInput(
                    config=ProviderConfig(
                        third_party_id="twitter",
                        clients=[
                            ProviderClientConfig(
                                client_id="4398792-WXpqVXRiazdRMGNJdEZIa3RVQXc6MTpjaQ",
                                client_secret="BivMbtwmcygbRLNQ0zk45yxvW246tnYnTFFq-LH39NwZMxFpdC"
                            )
                        ]
                    )
                )
            ]
        )
    ),
    multifactorauth.init(
        first_factors=["emailpassword", "thirdparty"],
        override=multifactorauth.OverrideConfig(
            functions=lambda original_implementation:
                override_multifactor_functions(original_implementation)
        )
    ),
    totp.init(),
    emailverification.init(
        mode="REQUIRED"
    ),
    accountlinking.init(
        should_do_automatic_account_linking=async_should_do_linking
    )
]

init(
    supertokens_config=supertokens_config,
    app_info=app_info,
    framework="django",
    recipe_list=recipe_list,
    mode="asgi",
    telemetry=False
)
"
`;

exports[`Python Template Generation > should generate correct config for ThirdParty only (Flask) 1`] = `
"from supertokens_python.recipe import session
from supertokens_python.recipe import dashboard
from supertokens_python.recipe import userroles
from supertokens_python.recipe import thirdparty
from supertokens_python.recipe.thirdparty.provider import ProviderInput, ProviderConfig, ProviderClientConfig



from supertokens_python import init, InputAppInfo, SupertokensConfig
import os

def get_api_domain() -> str:
    api_port = str(3001)
    api_url = f"http://localhost:{api_port}"
    return api_url

def get_website_domain() -> str:
    website_port = str(3000)
    website_url = f"http://localhost:{website_port}"
    return website_url

supertokens_config = SupertokensConfig(
    connection_uri="https://try.supertokens.com"
)

app_info = InputAppInfo(
    app_name="SuperTokens Demo App",
    api_domain=get_api_domain(),
    website_domain=get_website_domain(),
    api_base_path="/auth",
    website_base_path="/auth"
)

recipe_list = [
    session.init(),
    dashboard.init(),
    userroles.init(),
    thirdparty.init(
        sign_in_and_up_feature=thirdparty.SignInAndUpFeature(
            providers=[
                ProviderInput(
                    config=ProviderConfig(
                        third_party_id="google",
                        clients=[
                            ProviderClientConfig(
                                client_id="1060725074195-kmeum4crr01uirfl2op9kd5acmi9jutn.apps.googleusercontent.com",
                                client_secret="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"
                            )
                        ]
                    )
                ),
                ProviderInput(
                    config=ProviderConfig(
                        third_party_id="github",
                        clients=[
                            ProviderClientConfig(
                                client_id="467101b197249757c71f",
                                client_secret="e97051221f4b6426e8fe8d51486396703012f5bd"
                            )
                        ]
                    )
                ),
                ProviderInput(
                    config=ProviderConfig(
                        third_party_id="apple",
                        clients=[
                            ProviderClientConfig(
                                client_id="4398792-io.supertokens.example.service",
                                client_secret="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",
                                additional_config={
          "keyId": "7M48Y4RYDL",
          "privateKey": "-----BEGIN PRIVATE KEY-----\\nMIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgu8gXs+XYkqXD6Ala9Sf/iJXzhbwcoG5dMh1OonpdJUmgCgYIKoZIzj0DAQehRANCAASfrvlFbFCYqn3I2zeknYXLwtH30JuOKestDbSfZYxZNMqhF/OzdZFTV0zc5u5s3eN+oCWbnvl0hM+9IW0UlkdA\\n-----END PRIVATE KEY-----",
          "teamId": "YWQCXGJRJL"
}
                            )
                        ]
                    )
                ),
                ProviderInput(
                    config=ProviderConfig(
                        third_party_id="twitter",
                        clients=[
                            ProviderClientConfig(
                                client_id="4398792-WXpqVXRiazdRMGNJdEZIa3RVQXc6MTpjaQ",
                                client_secret="BivMbtwmcygbRLNQ0zk45yxvW246tnYnTFFq-LH39NwZMxFpdC"
                            )
                        ]
                    )
                )
            ]
        )
    )
]

init(
    supertokens_config=supertokens_config,
    app_info=app_info,
    framework="flask",
    recipe_list=recipe_list,
    mode="wsgi",
    telemetry=False
)
"
`;
